### PR TITLE
planner, expression: remove `TblID2Handle` from `Schema`

### PR DESCRIPTION
**Pull request by winoros**
### What problem does this PR solve? 

`TblID2Handle` can just be maintained when building plan. Only be held by just part of plan.
Not all of them need to hold it.
So remove it from `Schema` struct.

### What is changed and how it works?

Use a stack to maintain it when building plan. And then let speficific plan hold it.

### Check List 

Tests 

 - Unit test

Code changes

 - Has exported function/method change
 - Has exported variable/fields change

### CODE REVIEW

**Comment by winoros** [comment]
There're some small things can be splitted as small PRs. I'll do it later.

File: planner/core/logical_plan_builder.go [hunk 1]
```diff
@@ -2841,7 +2866,9 @@ func (b *PlanBuilder) buildDelete(delete *ast.DeleteStmt) (Plan, error) {
 		}
 	}
 
-	return del, nil
+	err = del.ResolveIndices()
```
Comment (lzmhhh123) [inline planner/core/logical_plan_builder.go]: Can we not do this?

File: planner/core/logical_plans.go [hunk 2]
```diff
@@ -493,7 +495,7 @@ func (ds *DataSource) deriveIndexPathStats(path *accessPath, conds []expression.
 	path.countAfterAccess = float64(ds.statisticTable.Count)
 	path.idxCols, path.idxColLens = expression.IndexInfo2Cols(ds.schema.Columns, path.index)
 	if !path.index.Unique && !path.index.Primary && len(path.index.Columns) == len(path.idxCols) {
-		handleCol := ds.getHandleCol()
+		handleCol := ds.getPKIsHandleCol()
```
Comment (lzmhhh123) [inline planner/core/logical_plans.go]: Why change it?

File: planner/core/logical_plan_builder.go [hunk 3]
```diff
@@ -2423,6 +2440,9 @@ func (b *PlanBuilder) projectVirtualColumns(ds *DataSource, columns []*table.Col
 	}
 
 	proj.SetSchema(ds.Schema().Clone())
+	for _, cols := range b.handleHelper.tailMap() {
+		cols[0] = proj.schema.RetrieveColumn(cols[0])
```
[the remark this answers is not in the document]
Comment (winoros) [reply planner/core/logical_plan_builder.go]: I didn't get you. 
The column in the map must be a cloned one though the unique id is not changed. It's something caused by the name changes. The loop can be removed after the names are removed from `Column`.

File: planner/core/logical_plan_builder.go (hunk 1 above)
Comment (winoros) [reply planner/core/logical_plan_builder.go]: No, we need to resolve the index for the handle column the delete executor holds.

File: planner/core/logical_plans.go (hunk 2 above)
Comment (winoros) [reply planner/core/logical_plans.go]: Oh, this two is nearly the same. I'll remove one of them.

File: executor/builder.go [hunk 4]
```diff
@@ -784,9 +785,7 @@ func (b *executorBuilder) buildUnionScanFromReader(reader Executor, v *plannerco
 	us := &UnionScanExec{baseExecutor: newBaseExecutor(b.ctx, v.Schema(), v.ExplainID(), reader)}
 	// Get the handle column index of the below plannercore.
 	// We can guarantee that there must be only one col in the map.
```
Comment (lamxTyler) [inline executor/builder.go]: Staled comment?

File: planner/core/point_get_plan.go [hunk 5]
```diff
@@ -555,3 +560,19 @@ func colInfoToColumn(db model.CIStr, tblName model.CIStr, asName model.CIStr, co
 		Index:       idx,
 	}
 }
+
+func (p *PointGetPlan) findHandleCol() *expression.Column {
+	// fields len is 0 for update and delete.
+	var handleCol *expression.Column
+	tbl := p.TblInfo
+	for i, col := range p.TblInfo.Columns {
+		if tbl.PKIsHandle && mysql.HasPriKeyFlag(col.Flag) {
```
Comment (lamxTyler) [inline planner/core/point_get_plan.go]: We can check the `tbl.PKIsHandle` first.

**Comment by eurekaka** [comment]
@winoros please resolve conflicts and fix CI.

File: planner/core/point_get_plan.go [hunk 6]
```diff
@@ -584,3 +577,21 @@ func colInfoToColumn(db model.CIStr, tblName model.CIStr, asName model.CIStr, co
 		Index:       idx,
 	}
 }
+
+func (p *PointGetPlan) findHandleCol() *expression.Column {
```
Comment (lzmhhh123) [inline planner/core/point_get_plan.go]: It's very same as `DataSource.getPKIsHandleCol`. Can we merge them?

File: planner/core/point_get_plan.go [hunk 7]
```diff
@@ -286,12 +292,9 @@ func checkFastPlanPrivilege(ctx sessionctx.Context, fastPlan *PointGetPlan, chec
...
 }
 
-func buildSchemaFromFields(ctx sessionctx.Context, dbName model.CIStr, tbl *model.TableInfo, fields []*ast.SelectField) *expression.Schema {
-	if dbName.L == "" {
-		dbName = model.NewCIStr(ctx.GetSessionVars().CurrentDB)
-	}
+func buildSchemaFromFields(dbName model.CIStr, tbl *model.TableInfo, fields []*ast.SelectField) *expression.Schema {
 	columns := make([]*expression.Column, 0, len(tbl.Columns)+1)
-	if len(fields) == 1 && fields[0].WildCard != nil {
+	if (len(fields) == 1 && fields[0].WildCard != nil) || len(fields) == 0 {
```
Comment (lamxTyler) [inline planner/core/point_get_plan.go]: When will `len(fields)` be 0? Why add the check？

**Comment by zz-jason** [comment]
@winoros Please fix CI to make this PR reviewable.

File: planner/core/logical_plan_builder.go [hunk 8]
```diff
@@ -890,6 +894,11 @@ func (b *PlanBuilder) buildUnion(ctx context.Context, union *ast.UnionStmt) (Log
 
 	oldLen := unionPlan.Schema().Len()
 
+	for i := 0; i < len(union.SelectList.Selects); i++ {
+		b.handleHelper.popMap()
```
Comment (lzmhhh123) [inline planner/core/logical_plan_builder.go]: If we pop all maps in this "clear handle node", then could it make these nodes not to push nil map?

File: executor/executor.go [hunk 9]
```diff
@@ -760,7 +762,8 @@ func (e *SelectLockExec) Open(ctx context.Context) error {
 	}
 
 	txnCtx := e.ctx.GetSessionVars().TxnCtx
-	for id := range e.Schema().TblID2Handle {
+	txnCtx.ForUpdate = true
```
[the remark this answers is not in the document]
Comment (winoros) [reply executor/executor.go]: This should be a merge error. I'll fix it in next commit.

File: planner/core/logical_plan_builder.go (hunk 8 above)
Comment (winoros) [reply planner/core/logical_plan_builder.go]: If we don't push `nil`.
We'll need this following codes I think:
```
func tail() {
	if tailPt == 0 {
		return nil
	}
	return stack[tailPt-1]
}
```
I think both of them is okay.

Comment (lzmhhh123) [reply planner/core/logical_plan_builder.go]: Yeah. All right. But don't you think the second one is more clear?

Comment (winoros) [reply planner/core/logical_plan_builder.go]: But union may be a subquery, the stack may not be nil before we enter the union.
Then don't push anything would be dangerous. We may pop wrong things out.

Comment (lzmhhh123) [reply planner/core/logical_plan_builder.go]: That's fine...

**Pull request merged**